### PR TITLE
Support customizing which KMS signing key to use.

### PR DESCRIPTION
**Pull request by steiza**
The default behavior is still to use HALLOW_KMS_KEY_ARN, but you can
customize the behavior of ChooseSigner() to use different KMS signing
keys based on IP, the user's identity, or some other aspect of their
request.

### CODE REVIEW

**Comment by alex** [comment]
Can you take a look at teh failing CI, thanks! I'll review afterwards

**Comment by paultag** [comment]
Looks like you may need a `gofmt`, @steiza !

File: main.go
Comment (paultag) [inline main.go:36]: I'd prefer this named something like `hallow.DefaultSigner` or something more explicit, like `hallow.KMSEnvSigner`, but that's not a huge one.

File: ca.go
Comment (paultag) [inline ca.go:32]: This should return err, not panic.

File: ca.go
Comment (paultag) [inline ca.go:37]: Same as above.

File: main.go
Comment (paultag) [reply main.go:36]: Alternatively this could be a function that allows passthrough, but again, this isn't a suggestion, just floating ideas. 

```
func PassthroughSigner(s crypto.Signer) func(events.APIGatewayProxyRequest) (crypto.Signer, error) {
    return func(events.APIGatewayProxyRequest) (crypto.Signer, error) {
        return s, nil
    }
}
```

File: main.go
Comment (paultag) [inline main.go:40]: You can just `return kmssigner.New...`

File: ca.go
Comment (paultag) [inline ca.go:30]: It may be worth having the default be sane here, `if s.ChooseSigner == nil` use `DefaultSigner` or something. That or check if nil and return an error.

**Review by paultag — CHANGES_REQUESTED** [review]
I know @alex was holding off on reviewing while tests were in flux, but since I think you're close, here's some initial feedback. I'm curious as to Alex's feedback too.

File: ca.go
Comment (alex) [inline ca.go:24]: I'd prefer if it instead of taking a func, this took an interface. That makes it much clearer to have state on the signerchooser.

An example of this is that the default one would be able to look up the env vars once on creation, and no on every signature.

File: ca.go
Comment (alex) [inline ca.go:29]: I think it'd make more sense for this to take a dedicated struct, instead of the whole event. That way the API surface is narrower and less prone to breaks.

Comment (steiza) [reply ca.go:33]: Oh, oops! I forgot to address this, let me work on an update.

File: ca.go
Comment (paultag) [inline ca.go:11]: I just realized this code doesn't have `golint` running as part of CI (that is an oversight) -- would you mind adding a comment to `SignerChooser` explaining what it is in a way that will cause `golint` to be happy?

File: main.go
Comment (paultag) [inline main.go:32]: Ditto for this comment, `golint` will trigger on this one - mind having it start with `APIGatewayContext contains some ...`

File: main.go
Comment (paultag) [inline main.go:39]: (same here, `DefaultSigner implements the ...`)

**Comment by paultag** [comment]
I resolved my comments about `golint` - our code drifted quite a bit, I'll send a MR after this is resolved

File: main.go
Comment (alex) [inline main.go:47]: I think we should reuse the `kmssigner` instance, constructing new AWS sessions is sloooow

**Comment by alex** [comment]
@paultag for context, their use case was somethign I dsicussed with them, and we came up with the interface idea as a way for them to use hallow for their use case without needing us to keep a bunch of code for it in hallow.

**Comment by paultag** [comment]
Yeah the approach seems fine :+1:

**Review by alex — APPROVED** [review]
This LGTM. I'll wait for @paultag to have a look before merging.

**Comment by paultag** [comment]
LGTM; I still have a nit about the `signerChooser` being nil, but it's not a big deal since it'll fail as a panic - I'll send a followup MR later. This is good to land

**Review by paultag — APPROVED** [review]

**Pull request merged**
